docs/decisions/0003-learning-path-enrollment-api.rst

```diff
@@ -0,0 +1,298 @@
++---------------------+-------------------------------------------------------+
+
+.. note::
+
+   The ``learning_path_id`` will be changed to a human readable content key.
+
+
+GET
+"""
+
+**Corresponding User Action:** Fetch enrollment in a Learning Path
+
+* When the ``username`` query parameter is present, the api returns
+
+  * the *active* enrollment object of the corresponding user for a **staff** user
+    or when both ``currentUser`` and the ``username`` point to the same user.
+  * HTTP 403 for a non-staff user, when ``currentUser`` and ``username``
+    refer to different users.
+
+* When the ``username`` query parameter is absent, the api returns
+
+  * all the *active* enrollments in a Learning Path if the ``currentUser`` is a **staff**
+    or **admin** user.
+  * the ``LearningPathEnrollment`` object if the logged-in user is enrolled in the
+    specified learning path. Otherwise returns a HTTP 404.
+
+
+POST
+""""
+
+**Corresponding User Action:** Enroll
+
+* When ``username`` is absent, enrolls the ``currentUser`` in the Learning
+  Path reference by the ``learning_path_id``.
+* When ``username`` is present
+
+  * ``username`` is different from the ``currentUser``, the user
+    linked to ``username`` will be enrolled, if the ``currentUser`` is a
+    **staff** or **admin**.
+  * Return an HTTP 403 error if the ``currentUser`` is a non-staff user
+  * Return an HTTP 404 if the ``username`` or ``learning_path_id`` do not
+    resolve into valid objects.
+
+* When an *active* ``LearningPathEnrollment`` object exists for the user and the
+  Learning Path, return HTTP 409 Conflict.
+
+
+DELETE
+""""""
+
+**Corresponding User Action:** Unenroll
+
+This will follow the same permission logic as the **POST** request allowing
+both self-unenrollment and staff-unenrollment.
+
+* Marks the *active* ``LearningPathEnrollment`` of the learner as inactive, by
+  setting the ``is_active`` field to ``False`` when the learner has an active
+  enrollment.
+* Return a 404 when there are no active enrollments for the learner.
+
+This method will be gated with a Django settings ``LEARNING_PATHS_ALLOW_SELF_UNENROLLMENT``
+and ``LEARNING_PATHS_ALLOW_STAFF_UNENROLLMENT`` that will default to ``False`` and
+will have to be explicitly enabled.
+
+* When the ``LEARNING_PATHS_ALLOW_STAFF_UNENROLLMENT`` is ``False``, the ``is_active``
+  field will be marked as **read-only**, preventing staff from unenrolling learners.
+* When the ``LEARNING_PATHS_ALLOW_SELF_UNENROLLMENT`` is ``False``, DELETE requests
+  from non-staff members will be returned HTTP 403.
+
+
+5. Fetch Enrollments API
+========================
+
+This API would list all the Learning Path enrollments
+
+* of the user making the request, for a non-staff user
+* of all users, for a staff user
+
++---------------------+-------------------------------------------------------+
+| API Path            | /api/v1/learning-path-enrollment/                     |
```

Comment:
Nit: this plugin [already has a separate URL namespace](https://github.com/open-craft/learning-paths-plugin/blob/2b40aaace2e27d276604d6523239151aa05f110c/learning_paths/apps.py#L24), so we can use `enrollment` instead of `learning-path-enrollment`.

Reply:
It looks the [regex is not defined](https://github.com/open-craft/learning-paths-plugin/blob/7584b126f99dea5cad3bceccbd69b9dde35917dd/learning_paths/apps.py#L27). So the urls are in the "global" namespace.

From the Django debug page
```
api/ v1/ ^programs/$ [name='learning-path-as-program-list']
api/ v1/ ^programs/(?P<pk>[^/.]+)/$ [name='learning-path-as-program-detail']
api/ v1/ learning-paths/<uuid:learning_path_uuid>/progress/ [name='learning-path-progress']
api/ v1/ learning-paths/<uuid:learning_path_uuid>/grade/ [name='learning-path-grade']
```

I noticed it while testing the learning-paths-as-programs functionality. I will add the regex and shorten the URL path in this one.